conftest.py

```diff
@@ -50,11 +96,36 @@ def pytest_collection_modifyitems(config, items):
             )
             item.add_marker(marker)
 
+    run_network_tests = environ.get('SKLEARN_SKIP_NETWORK_TESTS', '1') == '0'
+    skip_network = pytest.mark.skip(
+        reason="test requires internet connectivity")
+
+    # download datasets during collection to avoid thread unsafe behavior
+    # when running pytest in parallel with pytest-xdist
+    dataset_features_set = set(dataset_fetchers)
+    datasets_to_download = set()
+
+    for item in items:
+        item_keywords = set(item.keywords)
+        dataset_to_fetch = item_keywords & dataset_features_set
+        if not dataset_to_fetch:
+            continue
+
+        if run_network_tests:
+            datasets_to_download |= dataset_to_fetch
+        else:
+            # network tests are skipped
+            item.add_marker(skip_network)
+
+    # download datasets that are needed to avoid thread unsafe behavior
+    # by pytest-xdist
+    if run_network_tests:
+        for name in datasets_to_download:
+            dataset_fetchers[name]()
```

Comment:
Please add a verbose print statement here to make it easier to check that this is working as expected by looking at the CI logs:

```suggestion
            print("Sequential pre-caching of dataset {name}...")
            dataset_fetchers[name]()
```